{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initialization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from subprocess import call\n",
    "\n",
    "import pandas as pd\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "_FASTQ_DUMP_BIN = '/Users/dstone/software/sratoolkit.2.9.6-1-mac64/bin/fastq-dump' # dependent on your SRA installation\n",
    "def download_sra_fasta(sra_num: str, download_path: str=None, overwrite_existing: bool=False) -> int:\n",
    "    out_file = Path(download_path or './', f'{sra_num}.fasta').resolve()\n",
    "    if not overwrite_existing and out_file.exists():\n",
    "        print(f'{sra_num} already exists. Passing.')\n",
    "        return 0\n",
    "    \n",
    "    call_cmd = [_FASTQ_DUMP_BIN, '--fasta', sra_num]\n",
    "    if download_path is not None:\n",
    "          call_cmd.extend(['--outdir', str(Path(download_path).resolve())])\n",
    "    print(f'Downloading {sra_num} to fasta...')\n",
    "    ret_value = call(call_cmd)\n",
    "    if ret_value != 0:\n",
    "        raise Exception(f'Received return value from fastq-dump: {ret_value}')\n",
    "    print(f'Done!')\n",
    "    return ret_value\n",
    "\n",
    "_GENBANK_URL = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={acn}&rettype=fasta'\n",
    "def download_genbank_fasta(accession_num: str, download_path: str=None, overwrite_existing: bool=False) -> int:\n",
    "    out_file = Path(download_path or './', f'{accession_num}.fasta').resolve()\n",
    "    if not overwrite_existing and out_file.exists():\n",
    "        print(f'{accession_num} already exists. Passing.')\n",
    "        return 0\n",
    "    \n",
    "    call_cmd = ['wget', _GENBANK_URL.format(acn=accession_num)]\n",
    "    if download_path is not None:\n",
    "        call_cmd.extend(['--output-document', str(Path(download_path, f'{accession_num}.fasta').resolve())])\n",
    "    else:\n",
    "        call_cmd.extend(['--output-document', f'{accession_num}.fasta'])\n",
    "    print(f'Downloading {accession_num} to fasta...')\n",
    "    ret_value = call(call_cmd)\n",
    "    if ret_value != 0:\n",
    "        raise Exception(f'Received return value from wget: {ret_value}')\n",
    "    print(f'Done!')\n",
    "    return ret_value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download latest ncov sequence accession/SRA numbers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NCBI is maintaining a `.yaml` file that has Genbank accession + SRA run numbers, along with metadata, at the following url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-04-12 15:48:48--  https://www.ncbi.nlm.nih.gov/core/assets/genbank/files/ncov-sequences.yaml\n",
      "Resolving www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)... 130.14.29.110\n",
      "Connecting to www.ncbi.nlm.nih.gov (www.ncbi.nlm.nih.gov)|130.14.29.110|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 95960 (94K)\n",
      "Saving to: ‘/Users/dstone/projects/covid-19/ncov-sequences.yaml’\n",
      "\n",
      "/Users/dstone/proje 100%[===================>]  93.71K   389KB/s    in 0.2s    \n",
      "\n",
      "2020-04-12 15:48:49 (389 KB/s) - ‘/Users/dstone/projects/covid-19/ncov-sequences.yaml’ saved [95960/95960]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ncov_yaml_path = '/Users/dstone/projects/covid-19/ncov-sequences.yaml'\n",
    "!wget https://www.ncbi.nlm.nih.gov/core/assets/genbank/files/ncov-sequences.yaml \\\n",
    "    --output-document  $ncov_yaml_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now open that file\n",
    "with open(ncov_yaml_path, 'r') as stream:\n",
    "    ncov_data = yaml.safe_load(stream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'accession': 'MN908947',\n",
       " 'refseq-accession': 'NC_045512',\n",
       " 'gene-region': 'complete',\n",
       " 'collection-date': '2019-12',\n",
       " 'locality': {'country': 'China'}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncov_data['genbank-sequences'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MN908947 already exists. Passing.\n",
      "MN938384 already exists. Passing.\n",
      "MN975262 already exists. Passing.\n",
      "MN985325 already exists. Passing.\n",
      "MN988713 already exists. Passing.\n",
      "MN994467 already exists. Passing.\n",
      "MN994468 already exists. Passing.\n",
      "MN997409 already exists. Passing.\n",
      "MN988668 already exists. Passing.\n",
      "MN988669 already exists. Passing.\n",
      "MN996527 already exists. Passing.\n",
      "MN996528 already exists. Passing.\n",
      "MN996529 already exists. Passing.\n",
      "MN996530 already exists. Passing.\n",
      "MN996531 already exists. Passing.\n",
      "MT007544 already exists. Passing.\n",
      "LR757995 already exists. Passing.\n",
      "LR757996 already exists. Passing.\n",
      "LR757998 already exists. Passing.\n",
      "MT019529 already exists. Passing.\n",
      "MT019530 already exists. Passing.\n",
      "MT019531 already exists. Passing.\n",
      "MT019532 already exists. Passing.\n",
      "MT019533 already exists. Passing.\n",
      "MT020880 already exists. Passing.\n",
      "MT020881 already exists. Passing.\n",
      "MT027062 already exists. Passing.\n",
      "MT027063 already exists. Passing.\n",
      "MT027064 already exists. Passing.\n",
      "MT039873 already exists. Passing.\n",
      "MT039887 already exists. Passing.\n",
      "MT039888 already exists. Passing.\n",
      "MT039890 already exists. Passing.\n",
      "MT044257 already exists. Passing.\n",
      "MT044258 already exists. Passing.\n",
      "MT049951 already exists. Passing.\n",
      "MT066175 already exists. Passing.\n",
      "MT066176 already exists. Passing.\n",
      "MT072688 already exists. Passing.\n",
      "MT093571 already exists. Passing.\n",
      "MT093631 already exists. Passing.\n",
      "MT106052 already exists. Passing.\n",
      "MT106053 already exists. Passing.\n",
      "MT106054 already exists. Passing.\n",
      "MT118835 already exists. Passing.\n",
      "MT123290 already exists. Passing.\n",
      "MT123291 already exists. Passing.\n",
      "MT123292 already exists. Passing.\n",
      "MT123293 already exists. Passing.\n",
      "LC528232 already exists. Passing.\n",
      "LC528233 already exists. Passing.\n",
      "MT126808 already exists. Passing.\n",
      "MT135041 already exists. Passing.\n",
      "MT135042 already exists. Passing.\n",
      "MT135043 already exists. Passing.\n",
      "MT135044 already exists. Passing.\n",
      "MT152824 already exists. Passing.\n",
      "MT012098 already exists. Passing.\n",
      "MT050493 already exists. Passing.\n",
      "MT121215 already exists. Passing.\n",
      "MT159705 already exists. Passing.\n",
      "MT159706 already exists. Passing.\n",
      "MT159707 already exists. Passing.\n",
      "MT159708 already exists. Passing.\n",
      "MT159709 already exists. Passing.\n",
      "MT159710 already exists. Passing.\n",
      "MT159711 already exists. Passing.\n",
      "MT159712 already exists. Passing.\n",
      "MT159713 already exists. Passing.\n",
      "MT159714 already exists. Passing.\n",
      "MT159715 already exists. Passing.\n",
      "MT159716 already exists. Passing.\n",
      "MT159717 already exists. Passing.\n",
      "MT159718 already exists. Passing.\n",
      "MT159719 already exists. Passing.\n",
      "MT159720 already exists. Passing.\n",
      "MT159721 already exists. Passing.\n",
      "MT159722 already exists. Passing.\n",
      "MT066156 already exists. Passing.\n",
      "MT163716 already exists. Passing.\n",
      "MT163717 already exists. Passing.\n",
      "MT163718 already exists. Passing.\n",
      "MT163719 already exists. Passing.\n",
      "LC529905 already exists. Passing.\n",
      "MT184907 already exists. Passing.\n",
      "MT184908 already exists. Passing.\n",
      "MT184909 already exists. Passing.\n",
      "MT184910 already exists. Passing.\n",
      "MT184911 already exists. Passing.\n",
      "MT184912 already exists. Passing.\n",
      "MT184913 already exists. Passing.\n",
      "MT188339 already exists. Passing.\n",
      "MT188340 already exists. Passing.\n",
      "MT188341 already exists. Passing.\n",
      "MT192759 already exists. Passing.\n",
      "MT192765 already exists. Passing.\n",
      "MT192772 already exists. Passing.\n",
      "MT192773 already exists. Passing.\n",
      "MT198652 already exists. Passing.\n",
      "MT226610 already exists. Passing.\n",
      "MT233519 already exists. Passing.\n",
      "MT233522 already exists. Passing.\n",
      "MT233523 already exists. Passing.\n",
      "MT240479 already exists. Passing.\n",
      "MT246449 already exists. Passing.\n",
      "MT246450 already exists. Passing.\n",
      "MT246451 already exists. Passing.\n",
      "MT246452 already exists. Passing.\n",
      "MT246453 already exists. Passing.\n",
      "MT246454 already exists. Passing.\n",
      "MT246455 already exists. Passing.\n",
      "MT246456 already exists. Passing.\n",
      "MT246457 already exists. Passing.\n",
      "MT246458 already exists. Passing.\n",
      "MT246459 already exists. Passing.\n",
      "MT246460 already exists. Passing.\n",
      "MT246461 already exists. Passing.\n",
      "MT246462 already exists. Passing.\n",
      "MT246464 already exists. Passing.\n",
      "MT246466 already exists. Passing.\n",
      "MT246467 already exists. Passing.\n",
      "MT246468 already exists. Passing.\n",
      "MT246469 already exists. Passing.\n",
      "MT246470 already exists. Passing.\n",
      "MT246471 already exists. Passing.\n",
      "MT246472 already exists. Passing.\n",
      "MT246473 already exists. Passing.\n",
      "MT246474 already exists. Passing.\n",
      "MT246475 already exists. Passing.\n",
      "MT246476 already exists. Passing.\n",
      "MT246477 already exists. Passing.\n",
      "MT246478 already exists. Passing.\n",
      "MT246479 already exists. Passing.\n",
      "MT246480 already exists. Passing.\n",
      "MT246481 already exists. Passing.\n",
      "MT246482 already exists. Passing.\n",
      "MT246484 already exists. Passing.\n",
      "MT246485 already exists. Passing.\n",
      "MT246486 already exists. Passing.\n",
      "MT246487 already exists. Passing.\n",
      "MT246488 already exists. Passing.\n",
      "MT246489 already exists. Passing.\n",
      "MT246490 already exists. Passing.\n",
      "MT251972 already exists. Passing.\n",
      "MT251973 already exists. Passing.\n",
      "MT251974 already exists. Passing.\n",
      "MT251975 already exists. Passing.\n",
      "MT251976 already exists. Passing.\n",
      "MT251977 already exists. Passing.\n",
      "MT251978 already exists. Passing.\n",
      "MT251979 already exists. Passing.\n",
      "MT251980 already exists. Passing.\n",
      "MT253696 already exists. Passing.\n",
      "MT253697 already exists. Passing.\n",
      "MT253698 already exists. Passing.\n",
      "MT253699 already exists. Passing.\n",
      "MT253700 already exists. Passing.\n",
      "MT253701 already exists. Passing.\n",
      "MT253702 already exists. Passing.\n",
      "MT253703 already exists. Passing.\n",
      "MT253704 already exists. Passing.\n",
      "MT253705 already exists. Passing.\n",
      "MT253706 already exists. Passing.\n",
      "MT253707 already exists. Passing.\n",
      "MT253708 already exists. Passing.\n",
      "MT253709 already exists. Passing.\n",
      "MT253710 already exists. Passing.\n",
      "MT258377 already exists. Passing.\n",
      "MT258378 already exists. Passing.\n",
      "MT258379 already exists. Passing.\n",
      "MT258380 already exists. Passing.\n",
      "MT258381 already exists. Passing.\n",
      "MT258382 already exists. Passing.\n",
      "MT258383 already exists. Passing.\n",
      "MT259226 already exists. Passing.\n",
      "MT259227 already exists. Passing.\n",
      "MT259228 already exists. Passing.\n",
      "MT259229 already exists. Passing.\n",
      "MT259230 already exists. Passing.\n",
      "MT259231 already exists. Passing.\n",
      "MT259235 already exists. Passing.\n",
      "MT259236 already exists. Passing.\n",
      "MT259237 already exists. Passing.\n",
      "MT259239 already exists. Passing.\n",
      "MT259241 already exists. Passing.\n",
      "MT259243 already exists. Passing.\n",
      "MT259244 already exists. Passing.\n",
      "MT259245 already exists. Passing.\n",
      "MT259246 already exists. Passing.\n",
      "MT259247 already exists. Passing.\n",
      "MT259248 already exists. Passing.\n",
      "MT259249 already exists. Passing.\n",
      "MT259250 already exists. Passing.\n",
      "MT259251 already exists. Passing.\n",
      "MT259252 already exists. Passing.\n",
      "MT259253 already exists. Passing.\n",
      "MT259254 already exists. Passing.\n",
      "MT259256 already exists. Passing.\n",
      "MT259257 already exists. Passing.\n",
      "MT259258 already exists. Passing.\n",
      "MT259260 already exists. Passing.\n",
      "MT259261 already exists. Passing.\n",
      "MT259263 already exists. Passing.\n",
      "MT259264 already exists. Passing.\n",
      "MT259266 already exists. Passing.\n",
      "MT259267 already exists. Passing.\n",
      "MT259268 already exists. Passing.\n",
      "MT259269 already exists. Passing.\n",
      "MT259271 already exists. Passing.\n",
      "MT259273 already exists. Passing.\n",
      "MT259274 already exists. Passing.\n",
      "MT259275 already exists. Passing.\n",
      "MT259277 already exists. Passing.\n",
      "MT259278 already exists. Passing.\n",
      "MT259280 already exists. Passing.\n",
      "MT259281 already exists. Passing.\n",
      "MT259282 already exists. Passing.\n",
      "MT259284 already exists. Passing.\n",
      "MT259285 already exists. Passing.\n",
      "MT259286 already exists. Passing.\n",
      "MT259287 already exists. Passing.\n",
      "MT262896 already exists. Passing.\n",
      "MT262897 already exists. Passing.\n",
      "MT262898 already exists. Passing.\n",
      "MT262899 already exists. Passing.\n",
      "MT262900 already exists. Passing.\n",
      "MT262901 already exists. Passing.\n",
      "MT262902 already exists. Passing.\n",
      "MT262903 already exists. Passing.\n",
      "MT262904 already exists. Passing.\n",
      "MT262905 already exists. Passing.\n",
      "MT262906 already exists. Passing.\n",
      "MT262907 already exists. Passing.\n",
      "MT262908 already exists. Passing.\n",
      "MT262909 already exists. Passing.\n",
      "MT262910 already exists. Passing.\n",
      "MT262911 already exists. Passing.\n",
      "MT262912 already exists. Passing.\n",
      "MT262913 already exists. Passing.\n",
      "MT262914 already exists. Passing.\n",
      "MT262915 already exists. Passing.\n",
      "MT262916 already exists. Passing.\n",
      "MT262993 already exists. Passing.\n",
      "MT263074 already exists. Passing.\n",
      "MT263381 already exists. Passing.\n",
      "MT263382 already exists. Passing.\n",
      "MT263383 already exists. Passing.\n",
      "MT263384 already exists. Passing.\n",
      "MT263386 already exists. Passing.\n",
      "MT263387 already exists. Passing.\n",
      "MT263388 already exists. Passing.\n",
      "MT263390 already exists. Passing.\n",
      "MT263391 already exists. Passing.\n",
      "MT263392 already exists. Passing.\n",
      "MT263394 already exists. Passing.\n",
      "MT263395 already exists. Passing.\n",
      "MT263396 already exists. Passing.\n",
      "MT263398 already exists. Passing.\n",
      "MT263399 already exists. Passing.\n",
      "MT263400 already exists. Passing.\n",
      "MT263402 already exists. Passing.\n",
      "MT263403 already exists. Passing.\n",
      "MT263404 already exists. Passing.\n",
      "MT263405 already exists. Passing.\n",
      "MT263406 already exists. Passing.\n",
      "MT263408 already exists. Passing.\n",
      "MT263410 already exists. Passing.\n",
      "MT263411 already exists. Passing.\n",
      "MT263412 already exists. Passing.\n",
      "MT263413 already exists. Passing.\n",
      "MT263414 already exists. Passing.\n",
      "MT263415 already exists. Passing.\n",
      "MT263416 already exists. Passing.\n",
      "MT263417 already exists. Passing.\n",
      "MT263418 already exists. Passing.\n",
      "MT263419 already exists. Passing.\n",
      "MT263420 already exists. Passing.\n",
      "MT263421 already exists. Passing.\n",
      "MT263422 already exists. Passing.\n",
      "MT263423 already exists. Passing.\n",
      "MT263424 already exists. Passing.\n",
      "MT263425 already exists. Passing.\n",
      "MT263426 already exists. Passing.\n",
      "MT263428 already exists. Passing.\n",
      "MT263429 already exists. Passing.\n",
      "MT263430 already exists. Passing.\n",
      "MT263431 already exists. Passing.\n",
      "MT263432 already exists. Passing.\n",
      "MT263433 already exists. Passing.\n",
      "MT263434 already exists. Passing.\n",
      "MT263435 already exists. Passing.\n",
      "MT263436 already exists. Passing.\n",
      "MT263437 already exists. Passing.\n",
      "MT263438 already exists. Passing.\n",
      "MT263439 already exists. Passing.\n",
      "MT263440 already exists. Passing.\n",
      "MT263441 already exists. Passing.\n",
      "MT263442 already exists. Passing.\n",
      "MT263443 already exists. Passing.\n",
      "MT263444 already exists. Passing.\n",
      "MT263445 already exists. Passing.\n",
      "MT263446 already exists. Passing.\n",
      "MT263447 already exists. Passing.\n",
      "MT263448 already exists. Passing.\n",
      "MT263449 already exists. Passing.\n",
      "MT263450 already exists. Passing.\n",
      "MT263451 already exists. Passing.\n",
      "MT263452 already exists. Passing.\n",
      "MT263453 already exists. Passing.\n",
      "MT263454 already exists. Passing.\n",
      "MT263455 already exists. Passing.\n",
      "MT263456 already exists. Passing.\n",
      "MT263457 already exists. Passing.\n",
      "MT263458 already exists. Passing.\n",
      "MT263459 already exists. Passing.\n",
      "MT263462 already exists. Passing.\n",
      "MT263463 already exists. Passing.\n",
      "MT263464 already exists. Passing.\n",
      "MT263465 already exists. Passing.\n",
      "MT263467 already exists. Passing.\n",
      "MT263468 already exists. Passing.\n",
      "MT263469 already exists. Passing.\n",
      "LC534418 already exists. Passing.\n",
      "LC534419 already exists. Passing.\n",
      "Downloading MT233526 to fasta...\n",
      "Done!\n",
      "Downloading MT246667 to fasta...\n",
      "Done!\n",
      "Downloading MT276323 to fasta...\n",
      "Done!\n",
      "Downloading MT276324 to fasta...\n",
      "Done!\n",
      "Downloading MT276325 to fasta...\n",
      "Done!\n",
      "Downloading MT276326 to fasta...\n",
      "Done!\n",
      "Downloading MT276327 to fasta...\n",
      "Done!\n",
      "Downloading MT276328 to fasta...\n",
      "Done!\n",
      "Downloading MT276329 to fasta...\n",
      "Done!\n",
      "Downloading MT276330 to fasta...\n",
      "Done!\n",
      "Downloading MT276331 to fasta...\n",
      "Done!\n",
      "Downloading MT276597 to fasta...\n",
      "Done!\n",
      "Downloading MT276598 to fasta...\n",
      "Done!\n",
      "Downloading MT281577 to fasta...\n",
      "Done!\n",
      "Downloading MT291826 to fasta...\n",
      "Done!\n",
      "Downloading MT291827 to fasta...\n",
      "Done!\n",
      "Downloading MT291828 to fasta...\n",
      "Done!\n",
      "Downloading MT291829 to fasta...\n",
      "Done!\n",
      "Downloading MT291830 to fasta...\n",
      "Done!\n",
      "Downloading MT291831 to fasta...\n",
      "Done!\n",
      "Downloading MT291832 to fasta...\n",
      "Done!\n",
      "Downloading MT291833 to fasta...\n",
      "Done!\n",
      "Downloading MT291834 to fasta...\n",
      "Done!\n",
      "Downloading MT291835 to fasta...\n",
      "Done!\n",
      "Downloading MT291836 to fasta...\n",
      "Done!\n",
      "Downloading MT292569 to fasta...\n",
      "Done!\n",
      "Downloading MT292570 to fasta...\n",
      "Done!\n",
      "Downloading MT292571 to fasta...\n",
      "Done!\n",
      "Downloading MT292572 to fasta...\n",
      "Done!\n",
      "Downloading MT292573 to fasta...\n",
      "Done!\n",
      "Downloading MT292574 to fasta...\n",
      "Done!\n",
      "Downloading MT292575 to fasta...\n",
      "Done!\n",
      "Downloading MT293156 to fasta...\n",
      "Done!\n",
      "Downloading MT293158 to fasta...\n",
      "Done!\n",
      "Downloading MT293159 to fasta...\n",
      "Done!\n",
      "Downloading MT293160 to fasta...\n",
      "Done!\n",
      "Downloading MT293161 to fasta...\n",
      "Done!\n",
      "Downloading MT293162 to fasta...\n",
      "Done!\n",
      "Downloading MT293163 to fasta...\n",
      "Done!\n",
      "Downloading MT293164 to fasta...\n",
      "Done!\n",
      "Downloading MT293165 to fasta...\n",
      "Done!\n",
      "Downloading MT293166 to fasta...\n",
      "Done!\n",
      "Downloading MT293167 to fasta...\n",
      "Done!\n",
      "Downloading MT293168 to fasta...\n",
      "Done!\n",
      "Downloading MT293169 to fasta...\n",
      "Done!\n",
      "Downloading MT293170 to fasta...\n",
      "Done!\n",
      "Downloading MT293171 to fasta...\n",
      "Done!\n",
      "Downloading MT293172 to fasta...\n",
      "Done!\n",
      "Downloading MT293173 to fasta...\n",
      "Done!\n",
      "Downloading MT293174 to fasta...\n",
      "Done!\n",
      "Downloading MT293175 to fasta...\n",
      "Done!\n",
      "Downloading MT293176 to fasta...\n",
      "Done!\n",
      "Downloading MT293177 to fasta...\n",
      "Done!\n",
      "Downloading MT293178 to fasta...\n",
      "Done!\n",
      "Downloading MT293179 to fasta...\n",
      "Done!\n",
      "Downloading MT293180 to fasta...\n",
      "Done!\n",
      "Downloading MT293181 to fasta...\n",
      "Done!\n",
      "Downloading MT293182 to fasta...\n",
      "Done!\n",
      "Downloading MT293183 to fasta...\n",
      "Done!\n",
      "Downloading MT293184 to fasta...\n",
      "Done!\n",
      "Downloading MT293185 to fasta...\n",
      "Done!\n",
      "Downloading MT293186 to fasta...\n",
      "Done!\n",
      "Downloading MT293187 to fasta...\n",
      "Done!\n",
      "Downloading MT293188 to fasta...\n",
      "Done!\n",
      "Downloading MT293189 to fasta...\n",
      "Done!\n",
      "Downloading MT293190 to fasta...\n",
      "Done!\n",
      "Downloading MT293191 to fasta...\n",
      "Done!\n",
      "Downloading MT293192 to fasta...\n",
      "Done!\n",
      "Downloading MT293194 to fasta...\n",
      "Done!\n",
      "Downloading MT293195 to fasta...\n",
      "Done!\n",
      "Downloading MT293196 to fasta...\n",
      "Done!\n",
      "Downloading MT293197 to fasta...\n",
      "Done!\n",
      "Downloading MT293198 to fasta...\n",
      "Done!\n",
      "Downloading MT293199 to fasta...\n",
      "Done!\n",
      "Downloading MT293200 to fasta...\n",
      "Done!\n",
      "Downloading MT293201 to fasta...\n",
      "Done!\n",
      "Downloading MT293202 to fasta...\n",
      "Done!\n",
      "Downloading MT293204 to fasta...\n",
      "Done!\n",
      "Downloading MT293205 to fasta...\n",
      "Done!\n",
      "Downloading MT293206 to fasta...\n",
      "Done!\n",
      "Downloading MT293207 to fasta...\n",
      "Done!\n",
      "Downloading MT293208 to fasta...\n",
      "Done!\n",
      "Downloading MT293209 to fasta...\n",
      "Done!\n",
      "Downloading MT293210 to fasta...\n",
      "Done!\n",
      "Downloading MT293211 to fasta...\n",
      "Done!\n",
      "Downloading MT293212 to fasta...\n",
      "Done!\n",
      "Downloading MT293213 to fasta...\n",
      "Done!\n",
      "Downloading MT293215 to fasta...\n",
      "Done!\n",
      "Downloading MT293216 to fasta...\n",
      "Done!\n",
      "Downloading MT293218 to fasta...\n",
      "Done!\n",
      "Downloading MT293219 to fasta...\n",
      "Done!\n",
      "Downloading MT293220 to fasta...\n",
      "Done!\n",
      "Downloading MT293222 to fasta...\n",
      "Done!\n",
      "Downloading MT293224 to fasta...\n",
      "Done!\n",
      "Downloading MT293225 to fasta...\n",
      "Done!\n",
      "Downloading MT295464 to fasta...\n",
      "Done!\n",
      "Downloading MT295465 to fasta...\n",
      "Done!\n",
      "Downloading MT300186 to fasta...\n",
      "Done!\n",
      "Downloading MT304474 to fasta...\n",
      "Done!\n",
      "Downloading MT304475 to fasta...\n",
      "Done!\n",
      "Downloading MT304476 to fasta...\n",
      "Done!\n",
      "Downloading MT304477 to fasta...\n",
      "Done!\n",
      "Downloading MT304478 to fasta...\n",
      "Done!\n",
      "Downloading MT304479 to fasta...\n",
      "Done!\n",
      "Downloading MT304480 to fasta...\n",
      "Done!\n",
      "Downloading MT304481 to fasta...\n",
      "Done!\n",
      "Downloading MT304482 to fasta...\n",
      "Done!\n",
      "Downloading MT304483 to fasta...\n",
      "Done!\n",
      "Downloading MT304484 to fasta...\n",
      "Done!\n",
      "Downloading MT304485 to fasta...\n",
      "Done!\n",
      "Downloading MT304486 to fasta...\n",
      "Done!\n",
      "Downloading MT304487 to fasta...\n",
      "Done!\n",
      "Downloading MT304488 to fasta...\n",
      "Done!\n",
      "Downloading MT304489 to fasta...\n",
      "Done!\n",
      "Downloading MT304490 to fasta...\n",
      "Done!\n",
      "Downloading MT304491 to fasta...\n",
      "Done!\n",
      "Downloading MT308693, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308694, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308695, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308696, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308697, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308698, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308699, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308700, gapped to fasta...\n",
      "Done!\n",
      "Downloading MT308702 to fasta...\n",
      "Done!\n",
      "Downloading MT308703 to fasta...\n",
      "Done!\n",
      "Downloading MT308704 to fasta...\n",
      "Done!\n",
      "Downloading MT320538 to fasta...\n",
      "Done!\n",
      "Downloading MT320891 to fasta...\n",
      "Done!\n",
      "Downloading MT039874 to fasta...\n",
      "Done!\n",
      "Downloading MT077125 to fasta...\n",
      "Done!\n",
      "Downloading MT322394 to fasta...\n",
      "Done!\n",
      "Downloading MT322395 to fasta...\n",
      "Done!\n",
      "Downloading MT322396 to fasta...\n",
      "Done!\n",
      "Downloading MT322397 to fasta...\n",
      "Done!\n",
      "Downloading MT322398 to fasta...\n",
      "Done!\n",
      "Downloading MT322400 to fasta...\n",
      "Done!\n",
      "Downloading MT322401 to fasta...\n",
      "Done!\n",
      "Downloading MT322402 to fasta...\n",
      "Done!\n",
      "Downloading MT322403 to fasta...\n",
      "Done!\n",
      "Downloading MT322404 to fasta...\n",
      "Done!\n",
      "Downloading MT322405 to fasta...\n",
      "Done!\n",
      "Downloading MT322406 to fasta...\n",
      "Done!\n",
      "Downloading MT322407 to fasta...\n",
      "Done!\n",
      "Downloading MT322408 to fasta...\n",
      "Done!\n",
      "Downloading MT322409 to fasta...\n",
      "Done!\n",
      "Downloading MT322410 to fasta...\n",
      "Done!\n",
      "Downloading MT322411 to fasta...\n",
      "Done!\n",
      "Downloading MT322412 to fasta...\n",
      "Done!\n",
      "Downloading MT322413 to fasta...\n",
      "Done!\n",
      "Downloading MT322414 to fasta...\n",
      "Done!\n",
      "Downloading MT322415 to fasta...\n",
      "Done!\n",
      "Downloading MT322416 to fasta...\n",
      "Done!\n",
      "Downloading MT322417 to fasta...\n",
      "Done!\n",
      "Downloading MT322418 to fasta...\n",
      "Done!\n",
      "Downloading MT322419 to fasta...\n",
      "Done!\n",
      "Downloading MT322420 to fasta...\n",
      "Done!\n",
      "Downloading MT322421 to fasta...\n",
      "Done!\n",
      "Downloading MT322422 to fasta...\n",
      "Done!\n",
      "Downloading MT322423 to fasta...\n",
      "Done!\n",
      "Downloading MT322424 to fasta...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "# acquire all genbank data\n",
    "for genbank_dict in ncov_data['genbank-sequences']:\n",
    "    acn, completeness = genbank_dict['accession'], genbank_dict['gene-region']\n",
    "    if completeness == 'complete':\n",
    "        # only download if we have a full genome\n",
    "        download_genbank_fasta(acn, download_path='/Users/dstone/projects/covid-19/fastas/genbank/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this stalls after the first few downloads every time, so will stick with the genbank sequences for now\n",
    "# # acquire all SRA data\n",
    "# for genbank_dict in ncov_data['sra-accessions']:\n",
    "#     sra_num = genbank_dict['sra-run']\n",
    "#     download_sra_fasta(sra_num, download_path='/Users/dstone/projects/covid-19/fastas/sra/', overwrite_existing=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create dataframe of all accession information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'refseq-accession': 'NC_045512',\n",
       " 'gene-region': 'complete',\n",
       " 'collection-date': '2019-12',\n",
       " 'locality': {'country': 'China'}}"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncov_data['genbank-sequences'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_fields = ['collection-date', 'country', 'state']\n",
    "df_metadata = pd.DataFrame(columns=metadata_fields)\n",
    "for sd in ncov_data['genbank-sequences']:\n",
    "    row_dict = {}\n",
    "    for mf in metadata_fields:\n",
    "        if mf in ['country', 'state']:\n",
    "            row_dict[mf] = sd['locality'].get(mf)\n",
    "        else:\n",
    "            row_dict[mf] = sd[mf]\n",
    "    df_metadata.loc[sd['accession']] = row_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running analysis on sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run MUSCLE locally\n",
    "\n",
    "Get MUSCLE:\n",
    "```bash\n",
    " wget http://www.drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86darwin64.tar.gz --directory-prefix ~/projects/covid-19/\n",
    " cd ~/projects/covid-19/\n",
    " tar -xvzf muscle3.8.31_i86darwin64.tar.gz # yields a binary\n",
    " # combine above download fastas into one fasta for alignment\n",
    " cd ~/projects/covid-19/fastas/genbank\n",
    " # assumes there is one read per fasta (the above download code does yield this)\n",
    " cat *.fasta > all_ncov_genomes.fa\n",
    " # run MUSCLE, use default parameters\n",
    " cd ~/projects/covid-19/\n",
    " ./muscle3.8.31_i86darwin64 -in fastas/genbank/all_ncov_genomes.fa -out ./ncov_muscle_alignments.fa\n",
    " # time taken: ~\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Took too long (at least 24 hours). Current output:\n",
    "```\n",
    "dstone@5prime:~/projects/covid-19$ ./muscle3.8.31_i86darwin64 -in fastas/genbank/all_ncov_genomes.fa -out ./ncov_muscle_alignments.fa\n",
    "\n",
    "MUSCLE v3.8.31 by Robert C. Edgar\n",
    "\n",
    "http://www.drive5.com/muscle\n",
    "This software is donated to the public domain.\n",
    "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
    "\n",
    "all_ncov_genomes 485 seqs, max length 29945, avg  length 29847\n",
    "00:00:22     79 MB(0%)  Iter   1  100.00%  K-mer dist pass 1\n",
    "00:00:22     79 MB(0%)  Iter   1  100.00%  K-mer dist pass 2\n",
    "02:25:18  4706 MB(27%)  Iter   1  100.00%  Align node       \n",
    "02:25:18  4727 MB(28%)  Iter   1  100.00%  Root alignment\n",
    "04:26:24  4727 MB(28%)  Iter   2  100.00%  Refine tree   \n",
    "04:26:24  4727 MB(28%)  Iter   2  100.00%  Root alignment\n",
    "04:26:24  4727 MB(28%)  Iter   2  100.00%  Root alignment\n",
    "08:48:38  4727 MB(28%)  Iter   3  100.00%  Refine biparts\n",
    "^[[B^[[B  4727 MB(28%)  Iter   4   40.33%  Refine biparts\n",
    "```\n",
    "\n",
    "**Solution: look up if it is memory or CPU intensive and spin up an AWS ma"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build phylogenetic tree from MUSCLE output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alternate Method of building tree: k-mer based"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run jobs to compute k-mer profiles for each strain downloaded"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For varying values of `k`, use D_2 metric to compute distances"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build phylogenetic tree from this distance method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare with results from MUSCLE-alignment tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3_general",
   "language": "python",
   "name": "python3_general"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
